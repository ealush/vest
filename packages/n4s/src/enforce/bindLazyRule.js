import { RUN_RULE } from 'enforceKeywords';
import genRuleProxy from 'genRuleProxy';
import isFunction from 'isFunction';
import runtimeRules from 'runtimeRules';

// Initiates a chain of functions directly from the `enforce`
// function - that's even though we do not have any closure
// there to store that data.
export default function bindLazyRule(ruleName) {
  const registeredRules = [];

  const addFn = fnName => (...args) => {
    registeredRules.push(
      Object.defineProperty(
        value => runtimeRules[fnName](value, ...args),
        'name',
        { value: fnName }
      )
    );

    const returnvalue = genRuleProxy({}, addFn);

    return Object.assign(returnvalue, {
      [RUN_RULE]: getValue => {
        return registeredRules.every(fn => {
          // This  inversion of control when getting the value is
          // required in order to pass the function over to `shape`
          // so it can make the decision which args to pass to `optional`
          return fn(isFunction(getValue) ? getValue(fn.name) : getValue);
        });
      },
    });
  };

  return addFn(ruleName);
}
