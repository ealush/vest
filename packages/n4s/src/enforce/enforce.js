import compounds from 'compounds';
import { RUN_RULE } from 'enforceKeywords';
import runner from 'enforceRunner';
import isFunction from 'isFunction';
import isRule from 'isRule';
import proxySupported from 'proxySupported';
import rules from 'rules';

const rulesObject = Object.assign(rules(), compounds);

let rulesList = proxySupported() ? null : Object.keys(rulesObject);

const Enforce = value => {
  const proxy = genRuleProxy(enforce, ruleName => (...args) => {
    runner(rulesObject[ruleName], value, args);
    return proxy;
  });
  return proxy;
};

const enforce = genRuleProxy(Enforce, bindLazyRule);

enforce.extend = customRules => {
  Object.assign(rulesObject, customRules);

  if (!proxySupported()) {
    rulesList = Object.keys(rulesObject);
    genRuleProxy(Enforce, bindLazyRule);
  }

  return enforce;
};

export default enforce;

// Creates a proxy object that has access to all the rules
function genRuleProxy(target, output) {
  if (proxySupported()) {
    return new Proxy(target, {
      get: (target, fnName) => {
        if (isRule(rulesObject, fnName)) {
          return output(fnName);
        }

        return target[fnName];
      },
    });
  } else {
    /**
     * This method is REALLY not recommended as it is slow and iterates over
     * all the rules for each direct enforce reference. We only use it as a
     * lightweight alternative for the much faster proxy interface
     */
    return rulesList.reduce((target, fnName) => {
      return Object.defineProperties(target, {
        [fnName]: { get: () => output(fnName), configurable: true },
      });
    }, target);
  }
}

// Initiates a chain of functions directly from the `enforce`
// function - that's even though we do not have any closure
// there to store that data.
function bindLazyRule(ruleName) {
  const registeredRules = [];

  const addFn = fnName => (...args) => {
    registeredRules.push(
      Object.defineProperty(
        value => rulesObject[fnName](value, ...args),
        'name',
        { value: fnName }
      )
    );

    const returnvalue = genRuleProxy({}, addFn);

    return Object.assign(returnvalue, {
      [RUN_RULE]: getValue => {
        return registeredRules.every(fn => {
          // This  inversion of control when getting the value is
          // required in order to pass the function over to `shape`
          // so it can make the decision which args to pass to `optional`
          return fn(isFunction(getValue) ? getValue(fn.name) : getValue);
        });
      },
    });
  };

  return addFn(ruleName);
}
