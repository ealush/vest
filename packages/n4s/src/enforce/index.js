import { isRule, proxySupported } from '../lib';
import rules from '../rules';
import runner from './runner';

const rulesObject = { ...rules };

let enforce, rulesList;

if (proxySupported()) {
  enforce = value => {
    const proxy = new Proxy(rulesObject, {
      get: (rules, fnName) => {
        if (!isRule(rules, fnName)) {
          return;
        }

        return (...args) => {
          runner(rules[fnName], value, ...args);
          return proxy;
        };
      },
    });
    return proxy;
  };
} else {
  rulesList = Object.keys(rulesObject);

  enforce = value =>
    rulesList.reduce(
      (allRules, fnName) =>
        Object.assign(allRules, {
          ...(isRule(rulesObject, fnName) && {
            [fnName]: (...args) => {
              runner(rulesObject[fnName], value, ...args);
              return allRules;
            },
          }),
        }),
      {}
    );
}

enforce.extend = customRules => {
  Object.assign(rulesObject, customRules);

  if (!proxySupported()) {
    rulesList = Object.keys(rulesObject);
  }
};

export default enforce;
