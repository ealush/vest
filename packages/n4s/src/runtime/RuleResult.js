import hasOwnProperty from 'hasOwnProperty';

import { isBoolean } from 'isBoolean';
import { isEmpty } from 'isEmpty';
import { isNull } from 'isNull';
import { isUndefined } from 'isUndefined';
import { HAS_WARNINGS, HAS_ERRORS } from 'sharedKeys';

/**
 * Stores a rule result in an easy to inspect and manipulate structure.
 *
 * @param {boolean|RuleResult} ruleRunResult
 */
export default function RuleResult(ruleRunResult) {
  if (isUndefined(ruleRunResult)) {
    return;
  }

  if (isBoolean(ruleRunResult)) {
    this.setFailed(!ruleRunResult);
  } else {
    this.extend(ruleRunResult);
  }
}

/**
 * Determines whether a given value is a RuleResult instance
 * @param {*} res
 * @return {boolean}
 */
RuleResult.is = function (res) {
  return res instanceof RuleResult;
};

/**
 * Marks the current result object as an array
 */
RuleResult.prototype.asArray = function () {
  this.isArray = true;
  return this;
};

/**
 * @param {string} key
 * @param {value} value
 * @return {RuleResult} current instance
 */
RuleResult.prototype.setAttribute = function (key, value) {
  this[key] = value;
  return this;
};

/**
 * @param {boolean} failed
 * @return {RuleResult} current instance
 */
RuleResult.prototype.setFailed = function (failed) {
  this.setAttribute(this.warn ? HAS_WARNINGS : HAS_ERRORS, failed);
  return this.setAttribute('failed', failed);
};

/**
 * Adds a nested result object
 *
 * @param {string} key
 * @param {RuleResult} child
 */
RuleResult.prototype.setChild = function (key, child) {
  if (isNull(child)) {
    return null;
  }

  const isWarning =
    this[HAS_WARNINGS] || child[HAS_WARNINGS] || child.warn || this.warn;
  this.setAttribute(HAS_WARNINGS, (isWarning && child.failed) || false);
  this.setAttribute(
    HAS_ERRORS,
    this[HAS_ERRORS] ||
      child[HAS_ERRORS] ||
      (!isWarning && child.failed) ||
      false
  );
  this.setFailed(this.failed || child.failed);
  this.children = this.children || {};
  this.children[key] = child;
  return child;
};

/**
 * Retrieves a child by its key
 *
 * @param {string} key
 * @return {RuleResult|undefined}
 */
RuleResult.prototype.getChild = function (key) {
  return (this.children || {})[key];
};

/**
 * Extends current instance with a new provided result
 * @param {Boolean|RuleResult} newRes
 */
RuleResult.prototype.extend = function (newRes) {
  if (isNull(newRes)) {
    return this;
  }

  const res = RuleResult.is(newRes)
    ? newRes
    : new RuleResult().setAttribute('warn', !!this.warn).setFailed(!newRes);

  const failed = this.failed || res.failed;

  const children = mergeChildren(res, this).children;

  Object.assign(this, res);
  if (!isEmpty(children)) {
    this.children = children;
  }

  this.setFailed(failed);
  this.setAttribute(HAS_WARNINGS, !!(this[HAS_WARNINGS] || res[HAS_WARNINGS]));
  this.setAttribute(HAS_ERRORS, !!(this[HAS_ERRORS] || res[HAS_ERRORS]));
};

Object.defineProperty(RuleResult.prototype, 'pass', {
  get() {
    return !this.failed;
  },
});

/**
 * Deeply merge the nested children of compound rules
 *
 * @param {?RuleResult} base
 * @param {?RuleResult} add
 * @return {RuleResult}
 */
function mergeChildren(base, add) {
  const isRuleResultBase = RuleResult.is(base);
  const isRuleResultAdd = RuleResult.is(add);

  // If both base and add are result objects
  if (isRuleResultBase && isRuleResultAdd) {
    // Use failed if either is failing
    base.setFailed(base.failed || add.failed);

    // If neither has a children object, or the children object is
    if (isEmpty(base.children) && isEmpty(add.children)) {
      return base;
    }
    // If both have a children object
    if (base.children && add.children) {
      // Merge all the "right side" children back to base
      for (const key in base.children) {
        mergeChildren(base.children[key], add.children[key]);
      }

      // If a child exists in "add" but not in "base", just copy the child as is
      for (const key in add.children) {
        if (!hasOwnProperty(base.children, key)) {
          base.setChild(key, add.children[key]);
        }
      }

      // Return the modified base object
      return base;

      // If base has no children (but add does)
    } else if (!base.children) {
      // Use add's children
      base.children = add.children;

      // If add has no children
    } else if (!add.children) {
      // return base as is
      return base;
    }
    // If only base is `RuleResult`
  } else if (isRuleResultBase) {
    // Return base as is
    return base;

    // If only add is RuleResult
  } else if (isRuleResultAdd) {
    // Return add as is
    return add;
  }

  // That's a weird case. Let's fail. Very unlikely.
  return new RuleResult(false);
}
