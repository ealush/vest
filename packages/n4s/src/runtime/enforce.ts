import assign from 'assign';

import eachEnforceRule from 'eachEnforceRule';
import { ctx, TEnforceContext } from 'enforceContext';
import enforceEager, { TEnforceEager } from 'enforceEager';
import genEnforceLazy, { TLazyRules } from 'genEnforceLazy';
import isProxySupported from 'isProxySupported';
import { TRule, KBaseRules, baseRules, getRule } from 'runtimeRules';
/**
 * Enforce is quite complicated, I want to explain it in detail.
 * It is dynamic in nature, so a lot of proxy objects are involved.
 *
 * Enforce has two main interfaces
 * 1. eager
 * 2. lazy
 *
 * The eager interface is the most commonly used, and the easier to understand.
 * It throws an error when a rule is not satisfied.
 * The eager interface is declared in enforceEager.ts and it is quite simple to understand.
 * enforce is called with a value, and the return value is a proxy object that points back to all the rules.
 * When a rule is called, the value is mapped as its first argument, and if the rule passes, the same
 * proxy object is returned. Otherwise, an error is thrown.
 *
 * The lazy interface works quite differently. It is declared in genEnforceLazy.ts.
 * Rather than calling enforce directly, the lazy interface has all the rules as "methods" (only by proxy).
 * Calling the first function in the chain will initialize an array of calls. It stores the different rule calls
 * and the parameters passed to them. None of the rules are called yet.
 * The rules are only invoked in sequence once either of these chained functions are called:
 * 1. test(value)
 * 2. run(value)
 *
 * Calling run or test will call all the rules in sequence, with the difference that test will only return a boolean value,
 * while run will return an object with the validation result and an optional message created by the rule.
 */

function genEnforce(): TEnforce {
  const target = {
    context: () => ctx.useX(),
    extend: (customRules: TRule) => {
      assign(baseRules, customRules);
    },
  } as TEnforce;

  if (!isProxySupported()) {
    eachEnforceRule((ruleName: KBaseRules) => {
      // Only on the first rule access - start the chain of calls
      target[ruleName] = genEnforceLazy(ruleName);
    });

    return assign(enforceEager, target);
  }

  return new Proxy(assign(enforceEager, target) as TEnforce, {
    get: (target: TEnforce, key: string) => {
      if (key in target) {
        return target[key];
      }

      if (!getRule(key)) {
        return;
      }

      // Only on the first rule access - start the chain of calls
      return genEnforceLazy(key);
    },
  });
}

export const enforce = genEnforce();

type TEnforce = TEnforceMethods & TLazyRules & TEnforceEager;

type TEnforceMethods = {
  context: () => TEnforceContext;
  extend: (customRules: TRule) => void;
};
