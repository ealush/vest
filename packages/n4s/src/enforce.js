import bindLazyRule from 'bindLazyRule';
import bindExtend from 'enforce.extend';
import bindTemplate from 'enforce.template';
import runner from 'enforceRunner';
import genRuleProxy from 'genRuleProxy';
import runtimeRules from 'runtimeRules';
import withArgs from 'withArgs';

const Enforce = value => {
  const proxy = genRuleProxy({}, ruleName =>
    withArgs(args => {
      runner(runtimeRules[ruleName], value, args);
      return proxy;
    })
  );
  return proxy;
};

const enforce = genRuleProxy(Enforce, bindLazyRule);

bindExtend(enforce, Enforce);
bindTemplate(enforce);

export default enforce;
