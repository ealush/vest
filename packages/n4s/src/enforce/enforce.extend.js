import bindLazyRule from 'bindLazyRule';
import genRuleProxy from 'genRuleProxy';
import proxySupported from 'proxySupported';
import runtimeRules from 'runtimeRules';

export default function bindExtend(enforce, Enforce) {
  enforce.extend = customRules => {
    Object.assign(runtimeRules, customRules);

    if (!proxySupported()) {
      genRuleProxy(Enforce, bindLazyRule);
    }

    return enforce;
  };
}
