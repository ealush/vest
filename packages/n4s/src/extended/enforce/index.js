import enforce from '../../enforce';
import rules, {
  isAlphanumericLocales,
  isMobilePhoneLocales,
  isPostalCodeLocales,
} from '../rules';

enforce.extend(rules);

export default Object.assign(enforce, {
  isAlphanumericLocales,
  isMobilePhoneLocales,
  isPostalCodeLocales,
});
