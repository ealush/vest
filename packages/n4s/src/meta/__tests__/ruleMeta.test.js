import RuleResult from 'RuleResult';
import enforce from 'enforce';

describe('Rule with a message', () => {
  it('Should add message to result', () => {
    const res = enforce
      .shape({
        id: enforce.isNumeric().message('id must be numeric'),
        user: enforce.shape({
          name: enforce.shape({
            first: enforce
              .isString()
              .message(name => `${name} is not a valid first name`),
            last: enforce.isString().message('Last name is not valid'),
            middle: enforce
              .optional(enforce.isString())
              .message('When present, middle name must be a string'),
          }),
        }),
      })
      .run({
        user: {
          name: { first: 'somename', last: {} },
          id: 'id',
        },
      });

    expect(res.children.id.message).toBe('id must be numeric');
    expect(res.children.user.children.name.children.first.message).toBe(
      'somename is not a valid first name'
    );
    expect(res.children.user.children.name.children.middle.message).toBe(
      'When present, middle name must be a string'
    );
    expect(res).toMatchSnapshot();
  });
});

describe('Rule with a warning', () => {
  it('Should add a warning to the result', () => {
    const res = enforce
      .shape({
        id: enforce.isNumeric().warn(),
        user: enforce.shape({
          name: enforce.shape({
            first: enforce.isString().warn(),
            last: enforce.isString(),
            middle: enforce.optional(enforce.isString()).warn(),
            username: enforce.isString().longerThan(3).warn(),
          }),
        }),
      })
      .run({
        user: {
          name: { first: 'somename', last: {}, username: 'usr' },
          id: 'id',
        },
      });

    expect(res.children.id.warn).toBe(true);
    expect(res.children.user.children.name.children.first.warn).toBe(true);
    expect(res.children.user.children.name.children.middle.warn).toBe(true);
    expect(res).toMatchSnapshot();
  });
});

describe('Rule with "when"', () => {
  it('Should skip all excluded array items', () => {
    const Schema = enforce.shape({
      items: enforce.isArrayOf(
        enforce.shape({ checked: enforce.isBoolean() }).when(value => {
          return value.checked;
        })
      ),
    });

    const res = Schema.run({
      items: [
        { checked: true },
        { checked: false },
        { checked: false },
        { checked: true },
      ],
    });

    expect(res.children.items.children[0]).toBeInstanceOf(RuleResult);
    expect(res.children.items.children[1]).toBeUndefined();
    expect(res.children.items.children[2]).toBeUndefined();
    expect(res.children.items.children[3]).toBeInstanceOf(RuleResult);

    expect(res).toMatchSnapshot();
  });

  it('Should skip excluded object items', () => {
    const Schema = enforce.shape({
      example: enforce.isString().when(() => false),
      example1: enforce.isString(),
    });

    const res = Schema.run({
      example: 'something',
      example1: 'something',
    });

    expect(res.children.example).toBeUndefined();
    expect(res.children.example1).toBeInstanceOf(RuleResult);

    expect(res).toMatchSnapshot();
  });
  it('Should pass each item the value, key and parent', () => {
    const whenItems = jest.fn();
    const whenName = jest.fn();
    const Schema = enforce.shape({
      name: enforce
        .shape({
          first: enforce.isString(),
        })
        .when(whenName),
      items: enforce.isArrayOf(
        enforce.shape({ checked: enforce.isBoolean() }).when(whenItems)
      ),
    });

    const data = {
      name: {
        first: 'example',
      },
      items: [
        { checked: true },
        { checked: false },
        { checked: false },
        { checked: true },
      ],
    };

    Schema.run(data);

    expect(whenName).toHaveBeenCalledWith(data.name, 'name', data);

    expect(whenItems.mock.calls[0][0]).toBe(data.items[0]);
    expect(whenItems.mock.calls[1][0]).toBe(data.items[1]);
    expect(whenItems.mock.calls[2][0]).toBe(data.items[2]);
    expect(whenItems.mock.calls[3][0]).toBe(data.items[3]);
    expect(whenItems.mock.calls[0][1]).toBe(0); // key
    expect(whenItems.mock.calls[1][1]).toBe(1); // key
    expect(whenItems.mock.calls[2][1]).toBe(2); // key
    expect(whenItems.mock.calls[3][1]).toBe(3); // key
    expect(whenItems.mock.calls[0][2]).toBe(data.items); // parent
    expect(whenItems.mock.calls[1][2]).toBe(data.items); // parent
    expect(whenItems.mock.calls[2][2]).toBe(data.items); // parent
    expect(whenItems.mock.calls[3][2]).toBe(data.items); // parent
  });
});
