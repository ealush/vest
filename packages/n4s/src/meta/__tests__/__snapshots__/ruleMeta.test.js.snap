// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rule with "when" Should skip all excluded array items 1`] = `
RuleResult {
  "children": Object {
    "items": RuleResult {
      "children": Object {
        "0": RuleResult {
          "children": Object {
            "checked": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "warn": false,
            },
          },
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
          "when": false,
        },
        "3": RuleResult {
          "children": Object {
            "checked": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "warn": false,
            },
          },
          "failed": false,
          "hasErrors": false,
          "hasWarnings": false,
          "when": false,
        },
      },
      "failed": false,
      "hasErrors": false,
      "hasWarnings": false,
      "isArray": true,
    },
  },
  "failed": false,
  "hasErrors": false,
  "hasWarnings": false,
}
`;

exports[`Rule with "when" Should skip excluded object items 1`] = `
RuleResult {
  "children": Object {
    "example1": RuleResult {
      "failed": false,
      "hasErrors": false,
      "hasWarnings": false,
      "warn": false,
    },
  },
  "failed": false,
  "hasErrors": false,
  "hasWarnings": false,
}
`;

exports[`Rule with a message Should add message to result 1`] = `
RuleResult {
  "children": Object {
    "id": RuleResult {
      "failed": true,
      "hasErrors": true,
      "hasWarnings": false,
      "message": "id must be numeric",
      "warn": false,
    },
    "user": RuleResult {
      "children": Object {
        "name": RuleResult {
          "children": Object {
            "first": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "message": "somename is not a valid first name",
              "warn": false,
            },
            "last": RuleResult {
              "failed": true,
              "hasErrors": true,
              "hasWarnings": false,
              "message": "Last name is not valid",
              "warn": false,
            },
            "middle": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "message": "When present, middle name must be a string",
              "warn": false,
            },
          },
          "failed": true,
          "hasErrors": true,
          "hasWarnings": false,
        },
      },
      "failed": true,
      "hasErrors": true,
      "hasWarnings": false,
    },
  },
  "failed": true,
  "hasErrors": true,
  "hasWarnings": false,
}
`;

exports[`Rule with a warning Should add a warning to the result 1`] = `
RuleResult {
  "children": Object {
    "id": RuleResult {
      "failed": true,
      "hasErrors": false,
      "hasWarnings": true,
      "warn": true,
    },
    "user": RuleResult {
      "children": Object {
        "name": RuleResult {
          "children": Object {
            "first": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "warn": true,
            },
            "last": RuleResult {
              "failed": true,
              "hasErrors": true,
              "hasWarnings": false,
              "warn": false,
            },
            "middle": RuleResult {
              "failed": false,
              "hasErrors": false,
              "hasWarnings": false,
              "warn": true,
            },
            "username": RuleResult {
              "failed": true,
              "hasErrors": false,
              "hasWarnings": true,
              "warn": true,
            },
          },
          "failed": true,
          "hasErrors": true,
          "hasWarnings": true,
        },
      },
      "failed": true,
      "hasErrors": true,
      "hasWarnings": true,
    },
  },
  "failed": true,
  "hasErrors": true,
  "hasWarnings": true,
}
`;
